wsnet/cache.go

```diff
@@ -0,0 +1,154 @@
+package wsnet
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+)
+
+// dialerFunc is used to reference a dialer returned for caching.
+type dialerFunc func(ctx context.Context, key string) (*Dialer, error)
+
+// DialCache constructs a new DialerCache.
+// The cache clears connections that:
+// 1. Are older than the TTL and have no active user-created connections.
+// 2. Have been closed.
+func DialCache(ttl time.Duration, dialer dialerFunc) *DialerCache {
+	dc := &DialerCache{
+		ttl:         ttl,
+		dialerFunc:  dialer,
+		closed:      make(chan struct{}),
+		flightGroup: &singleflight.Group{},
+		mut:         sync.RWMutex{},
+		dialers:     make(map[string]*Dialer),
+		atime:       make(map[string]time.Time),
+	}
+	go dc.init()
+	return dc
+}
+
+type DialerCache struct {
+	dialerFunc  dialerFunc
+	ttl         time.Duration
+	flightGroup *singleflight.Group
+
+	closed  chan struct{}
+	mut     sync.RWMutex
+	dialers map[string]*Dialer
+	atime   map[string]time.Time
```

Comment:
You should use a struct instead of having two maps for the same key.

Reply:
I felt like it was a bit overkill here... I'll add a comment denoting they're the same.